src/types/quota.rs

```diff
@@ -0,0 +1,226 @@
+    /// The amount for that resource
+    pub amount: u64,
+}
+
+impl Display for QuotaResourceLimit<'_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{} {}", self.name, self.amount)
+    }
+}
+
+/// From [Resources](https://datatracker.ietf.org/doc/html/rfc2087#section-3)
+///
+/// Used by [`QuotaLimit`], and [`QuotaResource`]
+#[derive(Debug, Eq, PartialEq)]
+pub enum QuotaResourceName<'a> {
+    /// Sum of messages' RFC822.SIZE, in units of 1024 octets
+    Storage,
+    /// Number of messages
+    Message,
+    /// Any other string (for future RFCs)
+    Atom(Cow<'a, str>),
+}
+
+impl Display for QuotaResourceName<'_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Storage => write!(f, "STORAGE"),
+            Self::Message => write!(f, "MESSAGE"),
+            Self::Atom(s) => write!(f, "{}", s),
+        }
+    }
+}
+
+/// From [QUOTA Response](https://datatracker.ietf.org/doc/html/rfc2087#section-5.1)
+///
+/// Due to an implementation detail this is a wrapper around a single [`Quota`].
+///
+/// Used by [`Session::get_quota`] and [`Session::set_quota`]
+#[self_referencing]
+pub struct QuotaResponse {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    pub(crate) quota: Quota<'this>,
+}
+
+impl QuotaResponse {
+    /// Parse the [`Quota`] response from a response buffer.
+    pub fn parse(
+        owned: Vec<u8>,
+        unsolicited: &mut mpsc::Sender<UnsolicitedResponse>,
+    ) -> Result<Self, Error> {
+        QuotaResponseTryBuilder {
+            data: owned,
+            quota_builder: |input| {
+                let mut quotas = Vec::new();
+
+                // There should only be one single QUOTA response
+                parse_many_into(input, &mut quotas, unsolicited, |response| match response {
+                    Response::Quota(q) => Ok(MapOrNot::Map(Quota::from_imap_proto(q))),
+                    resp => Ok(MapOrNot::Not(resp)),
+                })?;
+
+                match quotas.len() {
+                    1 => Ok(quotas.remove(0)),
+                    _ => Err(Error::Parse(ParseError::Invalid(input.to_vec()))),
+                }
+            },
+        }
+        .try_build()
+    }
+
+    /// Access to the wrapped [`Quota`] struct
+    pub fn quota(&self) -> &Quota<'_> {
+        self.borrow_quota()
+    }
+}
+
+/// From [QUOTA Response](https://datatracker.ietf.org/doc/html/rfc2087#section-5.1)
+///
+/// Used by [`QuotaResponse`] and [`QuotaRoot`]
+#[derive(Debug, Eq, PartialEq)]
+pub struct Quota<'a> {
+    root_name: Cow<'a, str>,
+    resources: Vec<QuotaResource<'a>>,
+}
+
+impl<'a> Quota<'a> {
+    /// The quota root name
+    pub fn root_name(&self) -> &str {
+        &*self.root_name
+    }
+
+    /// The defined resources with their usage and limits (could be empty)
+    pub fn resources(&self) -> &[QuotaResource<'a>] {
+        &self.resources[..]
+    }
+}
```

Comment:
I'm torn about whether to keep these accessors or whether we should just make the fields of `Quota` `pub` and mark it as `#[non_exhaustive]` the same way we've done with other such types (like [`Mailbox`](https://docs.rs/imap/3.0.0-alpha.7/imap/types/struct.Mailbox.html)). What do you think?